docs/idl/NC-Framework.webidl

```diff
@@ -282,44 +284,63 @@ $macro(ModelDatatypes)
 	
 	enum NcDatatypeType {
-		"primitive",	// 0 primitive, e.g. NcUint16
-		"typedef",		// 1 typedef, i,e. simple alias of another datatype
-		"struct",		// 2 data structure
-		"enum",			// 3 enumeration
-		"null"			// 4 null
+		"Primitive",	// 0 primitive, e.g. NcUint16
+		"Typedef",		// 1 typedef, i,e. simple alias of another datatype
+		"Struct",		// 2 data structure
+		"Enum"
 	};
-	
+
 	interface NcDatatypeDescriptor {
 		NcName			name;	// datatype name
-		NcDatatypeType	type;	// primitive, typedef, struct, enum, or null
-		(NcString or NcName or sequence<NcFieldDescriptor> or sequence<NcEnumItemDescriptor> or null) content; // datatype content, see below
+		NcDatatypeType	type;	// Primitive, Typedef, Struct, Enum
 		
-		//  Contents of property 'content':
+		//  Further variants may introduce a content property as follows:
 		//
 		//		type		content
 		//	-----------------------------------------------------------------------------------------
-		//		primitive	empty string
-		//		typedef		name of referenced type
-		//		struct		sequence<NcFieldDescriptor>, one item per field of the struct	
-		// 	 	enum		sequence<NcEnumItemDescriptor>, one item per enum option
-		//		null		null
+		//		Primitive	no further content
+		//		Typedef		name of referenced type
+		//		Struct		sequence<NcFieldDescriptor>, one item per field of the struct
+		// 	 	Enum		sequence<NcEnumItemDescriptor>, one item per enum option
 		//	-----------------------------------------------------------------------------------------	
 	};
 	
+	interface NcDatatypeDescriptorPrimitive: NcDatatypeDescriptor {
+		//type will be Primitive
+	};
+
+	interface NcDatatypeDescriptorTypeDef: NcDatatypeDescriptor {
+		//type will be Typedef
+		NcName	content;	// original typedef datatype name
+	};
+
+	interface NcDatatypeDescriptorStruct: NcDatatypeDescriptor {
+		//type will be Struct
+		sequence<NcFieldDescriptor>	content;	// one item descriptor per field of the struct
+	};
+
+	interface NcDatatypeDescriptorEnum: NcDatatypeDescriptor {
+		//type will be Enum
+		sequence<NcEnumItemDescriptor>	content;	// one item descriptor per enum option
+	};
+	
 	// Descriptor of a class property
 	interface NcPropertyDescriptor {
-		NcPropertyId	id;			// element ID of property
-		NcName			name;		// name of property
-		NcName			typeName;	// name of property's datatype
-		NcBoolean		readOnly;	// TRUE iff property is read-only
-		NcBoolean		persistent;	// TRUE iff property value survives power-on reset
-		NcBoolean		required;	// TRUE iff property must be implemented
+		NcPropertyId	id;				// element ID of property
+		NcName			name;			// name of property
+		NcName			typeName;		// name of property's datatype. If empty then the type is any
+		NcBoolean		readOnly;		// TRUE iff property is read-only
+		NcBoolean		persistent;		// TRUE iff property value survives power-on reset
+		NcBoolean		required;		// TRUE iff property must be implemented
+		NcBoolean		isCollection;	// TRUE iff property is a collection
 	};
```

Comment:
For collection properties, I wonder if we need a field that gives the dimensionality.  For example, we'll need a 2D list datatype to describe the members of a matrix.  OCA has 2D lists and 2D arrays, for example.